pep-3157.txt

```diff
@@ -0,0 +1,247 @@
+PEP: 3157
+Title: Syntactic formalization of f-strings
+Version: $Revision$
+Last-Modified: $Date$
+Author: Pablo Galindo <pablogsal@python.org>,
+        Batuhan Taskaya <batuhan@python.org>,
+        Lysandros Nikolaou <lisandrosnik@gmail.com>
+Sponsor: Pablo Galindo <pablogsal@python.org>
+Discussions-To: python-dev@python.org
+Type: Standards Track
+Status: Draft
+Content-Type: text/x-rst
+Python-Version: 3.9
+Created: 15-Nov-2021
+
+
+Abstract
+========
+
+This document proposes to lift some of the restrictions originally formulated in
+:pep:`498` and to provide a formalized grammar for f-strings that can be
+integrated into parser directly. The proposed syntactic formalization of
+f-strings will have some small side-effects over how f-strings are parsed and
+interpreted, allowing for a considerable number of advantages for end users and
+library developers, while also improving dramatically the maintainance cost of
+the code dedicated to parse f-strings.
+
+
+Motivation
+==========
+
+When f-strings were originally introduced in :pep:`398` the specification was
+provided without providing a formal Grammar for f-strings. Additionally, theGG
+specification contains several restrictions that are imposed so the parsing of
+f-strings could be implemented into CPython without modifying the existing
+lexer. These limitations have been recognized previously and previous attempt
+have been made to lift them in :pep:`536` but non of this work was ever implemented 
+(see [2]_). Some of these limitations (collected originally by :pep:`536`) are:
+
+#. It is impossible to use the quote character delimiting the f-string
+   within the expression portion::
+
+    >>> f'Magic wand: { bag['wand'] }'
+                                 ^
+    SyntaxError: invalid syntax
+
+#. A previously considered way around it would lead to escape sequences
+   in executed code and is prohibited in f-strings::
+
+    >>> f'Magic wand { bag[\'wand\'] } string'
+    SyntaxError: f-string expression portion cannot include a backslash
+
+#. (TODO: DO WE WANT THIS?) Comments are forbidden even in multi-line f-strings::
```

Comment:
Ideally a similar example would be this:
 ```py
 # Valid
func(
    a = 1, # comment
)

# Invalid
func(
    a = 1, # comment)

# Valid
func(
    a = 1, # comment)
) # <======= the parentheses above is not included 
```

But I am not super sure about what happens when either the enclosing parentheses is found inside the comment or the enclosing string literals.

```py
# Valid
f'''A complex trick: {
    bag['bag']  # recursive bags!
}'''

# Completely invalid
f'''A complex trick: {
    bag['bag']  # recursive bags!}
'''

# Completely invalid
f'''A complex trick: {
    bag['bag']  # recursive bags!}'''


# Potentially valid
f'''A complex trick: {
    bag['bag']  # recursive bags!}
}''' # <=== paren is still here, because the one above is ignored

# Potentially valid
f'''A complex trick: {
    bag['bag']  # recursive bags!}'''
}''' # <=== the triple quotes above does not close the string, this one does
```

Should we decide on this?

Reply:
> when either the enclosing parentheses is found inside the comment or the enclosing string literals.

I think the more predictable way is that everything after the comment is just ignored, which means that if there is a "}" after the comment then is just ignored and part of the comment, which means that you still need an extra "}" as in the last example. Do you see any problems with this?

Reply:
agree with @pablogsal, ignoring everything after the `#` seems reasonable.

Reply:
agree with @pablogsal, ignoring everything after the `#` seems reasonable.